### DISCUSSION TITLE
Wrong placement of words in Arabic and Persian language!

### DISCUSSION THREAD

**Discussion by alefsoft1** [digao-dalpiaz/DzHTMLText#40] [General]
In Arabic and Persian language placement of words is wrong. For example "Hello World!" shows as "World! Hello";  
In Persian " سلام جهان" shows as "جهان سلام". I think Right-to-left mark(RLM) in the bi-directional text  is not considered.

**Comment by digao-dalpiaz** [comment]
Hello @alefsoft1

OK, in fact the component does not support Right-to-left text.
I gonna check if it's possible to add this functionality.

Can you include a new issue, please?
Thanks.

**Comment by sw-berrag** [comment]
Is there any implementation of  Right-to-left text?

**Reply by digao-dalpiaz** [reply]
Unfortunately, it is not supported yet. I need to study technically how RTL writing works to determine if it is possible to implement this in the component.

**Comment by sw-berrag** [comment]
Hello
I tried to find a way to make it possible.
So I made some modifications to the source code
Of course only in the VCL

```
private
.
.
.
FRealWidth: Integer;

published
.
.
.
{$ELSE}//VCL
    property BidiMode;
    property ParentBidiMode;
.
.
.


procedure TDzHTMLText.CanvasProcess(C: TCanvas);
var
  W: TDHVisualItem;
{$IFDEF FMX}
  RectDsn: TRectF;
  ColorDsn: TAlphaColor;
{$ENDIF}
begin
.
.
.
 {$IFDEF VCL}
  if IsRightToLeft then
  begin
   FRealWidth:=0;
    for W in VisualItems do
    begin
     FRealWidth:= Max(FRealWidth, W.Rect.Right);
    end;
    if not FAutoWidth then
    FRealWidth:= Max(FRealWidth, Width);
  end;
 {$ENDIF}
  for W in VisualItems do
    Paint_VisualItem(W, C);
end;

procedure TDzHTMLText.Paint_VisualItem(W: TDHVisualItem; C: TCanvas);
var
  R: TAnyRect;
begin
 {$IFDEF VCL}
   if IsRightToLeft then
    begin
     R.Top:= W.Rect.Top;
     R.Bottom:= W.Rect.Bottom;
     R.Left:= FRealWidth - W.Rect.Right;
     R.Right:= FRealWidth - W.Rect.Left;
    end
   else
 {$ENDIF}
  R := W.Rect;
.
.
.
end;
```

![Capture d'écran 2024-11-16 073858](https://github.com/user-attachments/assets/9f22a1ff-172d-448c-835f-88035490b474)

[https://i.ibb.co/jZR46WR/Video-2024-11-16-1.gif](url)

**Reply by digao-dalpiaz** [reply]
Ok, but from what I've researched about RTF writing, there are several details that change the position of the text.
For example, if there's a question mark, it will be inverted in relation to the entire sentence.
From what I understand, it's not just about aligning to the right, but rather it's necessary to parse the words in the text, to realign them according to a series of rules.

Or would it simply be aligning to the right?

**Reply by sw-berrag** [reply]
Yes, but the problem occurs with punctuation marks (?,:;...) when they are adjacent to the word. To avoid this confusion I modified this function:

procedure TDzHTMLText.Paint_Word(C: TCanvas; R: TAnyRect; W: TDHVisualItem_Word);
var
Flags: Longint;
begin
  R.Top := R.Top + W.YPos;

  {$IFDEF FMX}
  C.Fill.Color := C.Stroke.Color;
  C.FillText(R, W.Text, False, Opacity, [],
    TTextAlign.{$IF CompilerVersion >= 27}{XE6}Leading{$ELSE}taLeading{$ENDIF});
  {$ELSE}
  C.Brush.Style := bsClear;
    {$IFDEF MSWINDOWS}
    Flags:= DT_NOCLIP or DT_NOPREFIX;
    Flags:= DrawTextBiDiModeFlags(Flags);
    DrawTextW(C.Handle,
      PWideChar({$IFDEF FPC}UnicodeString(W.Text){$ELSE}W.Text{$ENDIF}),
      -1, R, Flags);
    {Using DrawText, because TextOut has no clip option, which causes
    bad overload of text when painting using background, oversizing the
    text area wildly.}
    {$ELSE}
    //FPC Linux
    C.TextOut(R.Left, R.Top, W.Text);
    {$ENDIF}
  {$ENDIF}
end;

**Reply by digao-dalpiaz** [reply]
Hello,

You may test the new feature using this branch: https://github.com/digao-dalpiaz/DzHTMLText/tree/rtl-text

**Comment by digao-dalpiaz** [comment]
The development of the RTL feature has started :)

Please, follow the steps in the issue #42

**Reply by sw-berrag** [reply]
![Capture d'écran 2024-12-23 184510](https://github.com/user-attachments/assets/ca06122a-b015-437b-97ca-b0ff2b39c55e)

![Capture d'écran 2024-12-23 185246](https://github.com/user-attachments/assets/ba38645c-918e-40f9-930e-b2ee0f930664)

**Reply by digao-dalpiaz** [reply]
The component already handles this elsewhere in the code, being generic for all platforms. This flag is specific to the Windows platform.

**Reply by digao-dalpiaz** [reply]
Just set new property "RightToLeftText" = True

**Reply by sw-berrag** [reply]
![Capture d'écran 2024-12-23 200232](https://github.com/user-attachments/assets/f9efc783-31c3-4ac1-a0d9-232a81f1bc4d)

**Reply by digao-dalpiaz** [reply]
Please write the text of the image here in the comment so I can reproduce the test here in the component. Thanks

**Reply by sw-berrag** [reply]
نص ملون، بالاخضر؟
<b>هذا نص مهم جدا</b> و <i>نص مائل</i>.<br><u>وأيضا 
لدينا خط مسطر</u> و 
<s>مشطوب</s> اذا اردت ذلك

[Text.txt](https://github.com/user-attachments/files/18232850/Text.txt)

**Reply by digao-dalpiaz** [reply]
Fixed. Please download the branch again and retest.

**Reply by sw-berrag** [reply]
![Capture d'écran 2024-12-23 221626](https://github.com/user-attachments/assets/ce936339-77e9-4258-b3e4-4078349ed673)

**Reply by digao-dalpiaz** [reply]
OK, now I've included the Arabic **semicolon** character as well.

**Reply by digao-dalpiaz** [reply]
Please, if you can, update your branch and test with the **semicolon** as well.

**Comment by sw-berrag** [comment]
Good.

But it will not be a general rule.
Because it is different with other languages ​​that are written from right to left such as:
Persian, Hebrew and Urdu..

**Comment by sw-berrag** [comment]
Bad formatting when text is currency in the Right to left text.
15,000.00  ---> 00.000,15
